keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+| CRD is deleted                                         | Nothing to type check against, but also means there are no coresponding custom resources |
+| CRD is in multiple clusters, but schema differs        | If policy author is aware of the schema variations, can they write policies that work for all the variations? |
+| Validation of an aggregated API server type            | Main API server does not have type definitions |
+
+Due to these complications, we have decided to evalute CEL expressions
+dynamically. Informational type checking will be provided (except for aggregated
+API server types), but will be surfaced only as warnings. See "Alternatives
+Considered" section for details of all the alternatives we reviewed when
+selecting this approach.
+
+Type checking is still performed for all expressions where a GVK can be matched
+to type check against, resulting in warnings, e.g.:
+
+```yaml
+...
+status:
+  expressionWarnings:
+    - expression: "object.foo"
+      warning: "no such field 'foo'"
+```
+
+#### Failure Policy
+
+For in-process validation there is no remote request, so errors should be
+deterministic. We believe this significantly reduces the risk of "fail closed"
+admission control as compared to webhooks.
+
+Because failure policy is most often selected based on the need to guarantee
+enforcement, we will default failure policy to "fail" and allow it to
+be configured on a per-policy basis:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+spec:
+  ...
+  failurePolicy: Ignore # The default is "Fail"
+  validations:
+    - expression: "object.spec.xyz == params.x"  
+```
+
+#### Safety measures
+
+To prevent clusters from being put into a unustable state that cannot be
+recoverd from via the API, admission webhooks are not allowed to match
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.
+
+We will extend this approach:
+
+- `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`
+- `ValidatingWebhookConfiguration` cannot match `ValidatingAdmissionPolicy` or
+  `MutatingWebhookConfiguration`.
+
+Note that this does allow `ValidatingAdmissionPolicy` to match
+`ValidatingWebhookConfiguration`.
+
+Note: In the future we may further loosen this up and allow admission
+configuration to intercept/guard writes to admission configuration while
+preventing deadlock - Add feature to configure a set of webhooks to intercept
+other webhooks https://github.com/kubernetes/kubernetes/issues/101794.
+
+Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
+`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
+higher level. This could be added later.
+
+#### Default Validations
+
+A `defaultValidations` field will allow policy definition authors to define
+expressions which are evaluated when the policy definition matches a resource
+but no policy configurations match that resource.
```

Comment:
what happens if a multiple param objects are bound, or one param object is bound several times (via separate bindings that match in different ways)?

Reply:
Ooh, I didn't rethink this after splitting binding and config. We don't need to evaluate the CEL expressions multiple times for the same params since the result will always be the same. But the bindings may have different enforcement settings, so we will need to apply those correctly and attribute all violations to the appropriate binding. I'll add an implementation note.